### DISCUSSION TITLE
Child, but single row?

### DISCUSSION THREAD

**Discussion by amyworrall** [ricomariani/CG-SQL-author#116] [Q&A] [answered]
I've been making extensive use of child result sets. In my Swift codegen, they neatly appear as an array of child objects.

However, I'm wondering if a result can have a child that's a single row, such as is produced by a sproc that uses `OUT` to return only a single row.

Why would I want such a thing, as opposed to just integrating the relevant columns into my existing output row? Essentially to keep things separate, and to make it so that column name collisions don't matter.

The only way I can think of doing this at the moment is somehow wrapping things in a custom `OBJECT<>` type (since `OBJECT<_ SET>` refers only to result sets, I believe). If I'm not mistaken, there's no built in support to help me do this -- am I right? Has anyone done something similar?

**Comment by ricomariani** [comment]
"OUT" also returns a set... It's just a single row result set.  The interface is different only because the row number is not needed.  But OUT can be 0 or 1 rows so even count is needed.  Underneath it all, it's just a rowset same as all the others.

You can also just "OUT UNION" one row.

________________________________
From: amyworrall ***@***.***>
Sent: Sunday, November 10, 2024 7:40 AM
To: ricomariani/CG-SQL-author ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ricomariani/CG-SQL-author] Child, but single row? (Discussion #116)


I've been making extensive use of child result sets. In my Swift codegen, they neatly appear as an array of child objects.

However, I'm wondering if a result can have a child that's a single row, such as is produced by a sproc that uses OUT to return only a single row.

Why would I want such a thing, as opposed to just integrating the relevant columns into my existing output row? Essentially to keep things separate, and to make it so that column name collisions don't matter.

The only way I can think of doing this at the moment is somehow wrapping things in a custom OBJECT<> type (since OBJECT<_ SET> refers only to result sets, I believe). If I'm not mistaken, there's no built in support to help me do this -- am I right? Has anyone done something similar?

**Comment by ricomariani** [comment]
Both of these work for instance:

```
proc foo()
begin
  declare C cursor like(x int!, y text!);
  fetch C using 1 x, "2" y;
  out C;
end;


proc bar1()
begin
  -- this will be a result set object of type object<foo set>;
  let r := foo();
end;

proc bar2()
begin
  -- this is the same only the declaration isn't implied
  var u object<foo set>;
  set u := foo();
end;
```
________________________________
From: amyworrall ***@***.***>
Sent: Sunday, November 10, 2024 7:40 AM
To: ricomariani/CG-SQL-author ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ricomariani/CG-SQL-author] Child, but single row? (Discussion #116)


I've been making extensive use of child result sets. In my Swift codegen, they neatly appear as an array of child objects.

However, I'm wondering if a result can have a child that's a single row, such as is produced by a sproc that uses OUT to return only a single row.

Why would I want such a thing, as opposed to just integrating the relevant columns into my existing output row? Essentially to keep things separate, and to make it so that column name collisions don't matter.

The only way I can think of doing this at the moment is somehow wrapping things in a custom OBJECT<> type (since OBJECT<_ SET> refers only to result sets, I believe). If I'm not mistaken, there's no built in support to help me do this -- am I right? Has anyone done something similar?

**Comment by ricomariani** [comment, marked as answer]
This is a complete example in Lua:

```
declare proc printf no check;

proc foo()
begin
  cursor C like(x int!, y text!);
  fetch C using 100 x, "str" y;
  out C;
end;

proc bar1()
begin
  -- this will be a result set object of type object<foo set>;
  let r := foo();

  cursor O like(a int, b int, child object<foo set>);
  fetch O using
     1 a,  2 b, r child;

  out O;
end;

proc bar2()
begin
   cursor C fetch from call bar1();
   printf("%d %d: ", C.a, C.b);
   cursor D for C.child;
   fetch D;
   printf("%d %s\n", D.x, D.y);
end;

— you can do this in C you just need to echo a main or otherwise get a main
@echo lua, "bar2()";

$ lua x.l
1 2: 100 str
```

________________________________
From: amyworrall ***@***.***>
Sent: Sunday, November 10, 2024 7:40 AM
To: ricomariani/CG-SQL-author ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ricomariani/CG-SQL-author] Child, but single row? (Discussion #116)


I've been making extensive use of child result sets. In my Swift codegen, they neatly appear as an array of child objects.

However, I'm wondering if a result can have a child that's a single row, such as is produced by a sproc that uses OUT to return only a single row.

Why would I want such a thing, as opposed to just integrating the relevant columns into my existing output row? Essentially to keep things separate, and to make it so that column name collisions don't matter.

The only way I can think of doing this at the moment is somehow wrapping things in a custom OBJECT<> type (since OBJECT<_ SET> refers only to result sets, I believe). If I'm not mistaken, there's no built in support to help me do this -- am I right? Has anyone done something similar?

**Reply by ricomariani** [reply]
For a C target:

```
@echo c, '

// a database is not required for this trivial app
int main(int argc, char **argv) {
  // normally a database but I am lazy today
  if (bar2(NULL)) printf("error\n");
}

';

$ cql foo.sql --cg foo.h foo.c
$ cc foo.c cqlrt.c -lsqlite3
$ ./a.out
1 2: 100 str
```

**Comment by ricomariani** [comment]
Note:  The whole point of the helper 

```
OUT UNION call foo() JOIN call bar() using (id);
```

Is to *not* call `bar` one time for each row but rather get all the results and then join them afterwards.

If you only have a procedure that does single rows you have to do the nested loops thing yourself but it's pretty easy:

```
declare proc printf no check;

create table my_data(
 id int!,
 name text!,
 age int!
);

proc foo()
begin
  select id, name from my_data;
end;

proc bar(id int!)
begin
  cursor C like (id int, something int);
  fetch C using  id id,  id*id something;
  out C;
end;

proc foo_and_bar()
begin
  cursor C for call foo();
  loop fetch C
  begin
     -- make a cursor like the one foo returns but with one more column
     cursor D like (like foo, child object<bar set>);
     
     -- load it up with the existing values and a call to bar
     -- note: this works even if bar returns many rows
     fetch D from values(from C, bar(C.id));
     
     -- emit it
     out union D;
  end;
end;
```

**Comment by amyworrall** [comment]
So in your example in the post above,

let's assume that `bar(id int!)` can return either 0 or 1 rows.

Inside the code for `foo_and_bar()`, how can I distinguish whether the `child` field (i.e. the fetch from bar) contains 0 or 1 rows? Suppose I only want to `out union D` if `bar()` returned a row, for example… how would I do that?

I assume that `D.child` is always not null, because there's always a child result set, even if it has no items in it…

**Comment by ricomariani** [comment]
```sql
declare proc printf no check;

proc make_schema()
begin
  create table my_data(
   id int!,
   name text!,
   age int!
  );
end;

proc foo()
begin
  select id, name from my_data;
end;

proc bar(id int!)
begin
  if id % 2 then
    cursor C like (id int, something int);
    fetch C using  id id,  id*id something;
    -- if you only do this on some paths the other paths get an empty result for free
    out C;  
  end;
end;

proc use_foo_and_bar()
begin
  cursor C for call foo_and_bar();
  loop fetch C
  begin
     printf("%d, %s\n", C.id, C.name);
     cursor D for C.child;
     fetch D;  -- this could have been loop fetch D but since it's only 0 or 1 rows...
     if D then
        printf(" --> child --> %d, %d\n", D.id, D.something);
     end;
  end;
end;

proc go()
begin
  make_schema();
  let i := 0;
  while i < 5
  begin
    i += 1;
    insert into my_data() values() @dummy_seed(i);
  end;
  use_foo_and_bar();
end;

@echo lua, "go(sqlite3.open_memory())\n";
```

This generates:

```
1, name_1
 --> child --> 1, 1
2, name_2
3, name_3
 --> child --> 3, 9
4, name_4
5, name_5
 --> child --> 5, 25
 ```

**Comment by ricomariani** [comment]
Do you cqlrt_cf for your Swift interop?

**Reply by amyworrall** [reply]
Yes I do.

My Swift codegen layer is [here](https://gist.github.com/amyworrall/888eea17a7d0466d4bdcfaf2d25d495e). Note that it adds a bunch of features specific to my project, so it's not a drop-in reusable one. Also I'm not quite on the latest CGSQL so a few things will need updating!

Here's how I trigger CQL:

```
cc -E -x c -Wno-invalid-pp-token $INPUT_FILE_NAME | ${PROJECT_DIR}/../CQL/cql  --cqlrt cqlrt_cf.h --cg "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.cql.h" "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.cql.c" --generate_type_getters

cc -E -x c -Wno-invalid-pp-token $INPUT_FILE_NAME | ${PROJECT_DIR}/../CQL/cql --rt json_schema --cqlrt cqlrt_cf.h --cg "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.json" --generate_type_getters

cc -E -x c -Wno-invalid-pp-token $INPUT_FILE_NAME | ${PROJECT_DIR}/../CQL/cql --rt objc_mit --objc_c_include_path $INPUT_FILE_BASE.cql.h --cqlrt cqlrt_cf.h --cg "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.objc.h" --generate_type_getters
```

Then, my codegen adds a few layers:

* My `Filename.helper.objc.h`/`.m` files are so that I can consume CGSQL's objc codegenned classes in an objc class of my own. I found that trying to use them directly from Swift was causing me headaches with regard to memory management. It may have been resolvable another way but this works for me.
* My `Filename.swift` provide a Swift interface (based on pure Swift structs), and iterate through every row in the CGSQL result set to initialise a Swift struct from them. This does mean that all the rows are read at the point a query is performed, but for a client app like mine, I'd prefer to err on the side of developer productivity! (This is the same tradeoff we made with Work Chat inside Meta.)

**Reply by ricomariani** [reply]
The reason I asked is that I can't built cqlrt_cf myself anymore I don't have a Mac and Objective C is highly resistant to any compilation on Ubuntu no matter how hard I try...  I wanted to make sure it was still working.

* I deleted vault_sensitive it was big and it seemed like nobody was using it, it's apparently only used on iOS on Meta (I did that yesterday)
* I added macros and include pre-processing some time ago (like 2 years?)  so this kind of stuff `cc -E -x c -Wno-invalid-pp-token $INPUT_FILE_NAME` shouldn't be necessary, I'm not sure it even works anymore.  I recommend you switch to @macro(macro_type) as soon as you can :D -- CQL macros are very powerful and they don't need crazy escaping and stuff.
* I want to delete `--rt objc` and replace it with python that reads `--rt JSON` like we did at Meta to make the output.  I think I can make the output identical and maybe even add streaming getters so you don't have to materialize the result set (only to throw it away) if you are storing it yourself.  Again, this should be very easy, but I can't test it.... so, I was thinking maybe I could add the generator and get some Mac users to try it out before deleting the C version.  The python version would be much more approachable to customize also.
* I believe `--generate_type_getters` doesn't work anymore, well, you don't need the option.  The non-type-getters version of the code had serious errors and it was deleted some time ago so the type getters flavor is the only flavor now.  Again, this let me delete a bunch of code from the compiler.  This happened about a year ago also.

I don't think it would be much work for you to get on the latest keeping the above in mind.  Unless I broke --objc along the way but if I did I would appreciate a heads up on that, I'll fix it for you.

The easiest way to try the latest is to make a fork and build the amalgam.  Or drop in the binaries from the fork if that's doable.

**Reply by amyworrall** [reply]
I've been meaning to migrate my project (Deluge) to latest, it's just a case of time.

I have tried building latest CGSQL within the last month, and it was working then. I just didn't get as far as integrating it into Deluge. That's not since the removal of `vault_sensitive` of course, but I'm not using that feature so can't imagine it's a problem!

> maybe I could add the generator and get some Mac users to try it out before deleting the C version

I am happy to try things out!

> I believe --generate_type_getters doesn't work anymore, well, you don't need the option

I think that's true. When I last played with a later version of cql I did not need that option.

**Comment by ricomariani** [comment]
I threw together https://github.com/ricomariani/CG-SQL-author/blob/main/sources/cqlrt_cf/cqlobjc.py which is a complete replacement for `--rt objc` (it makes the exact same output except possibly in a different order)

You should be able to use it with any CQL from the last several years.

**Comment by ricomariani** [comment]
I changed ./make.sh in the cqlrt_cf directory to illustrate the way to use the python script, it's pretty apparent from the usage but a demo never hurts.  I can't actually test it other than to verify that it makes the same source code.  I can't run ObjectiveC with Framework (I can without but that proves nothing).

**Comment by ricomariani** [comment]
I have this PR set to remove --rt objc completely

https://github.com/ricomariani/CG-SQL-author/pull/154

**Comment by ricomariani** [comment]
There is more discussion of the new stuff here:  https://github.com/ricomariani/CG-SQL-author/discussions/155

If you can try it out at some point I would appreciate it!

**Reply by amyworrall** [reply]
Hi. Not had time for a deep dive, but just to check: If I download the current source from the repo, use `make clean` and `make` to build CQL, then go to `sources/objc_full` and run `clean.sh` and `make.sh` -- that should be sufficient to test the new stuff?

I tried that and I'm afraid it's giving type mismatches to the local CoreFoundation:

```
amyworrall@Mac objc_full % ./make.sh 
building cql
make: Nothing to be done for `all'.
This script minimally tests the generated code to see if it will at least compile
building C code
building JSON
building OBJC header (.h)
building OBJC implementation (.m)
compiling generated code looking for errors
In file included from Sample_objc.m:2:
In file included from ./Sample_objc.h:6:
In file included from ./Sample.h:4:
../cqlrt_cf/cqlrt_cf.h:24:15: error: typedef redefinition with different types ('void *' vs 'const void *')
   24 | typedef void *CFTypeRef;
      |               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFBase.h:508:42: note: previous definition is here
  508 | typedef const CF_BRIDGED_TYPE(id) void * CFTypeRef;
      |                                          ^
In file included from Sample_objc.m:2:
In file included from ./Sample_objc.h:6:
In file included from ./Sample.h:4:
../cqlrt_cf/cqlrt_cf.h:25:15: error: typedef redefinition with different types ('void *' vs 'const struct __CFData *')
   25 | typedef void *CFDataRef;
      |               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFData.h:18:57: note: previous definition is here
   18 | typedef const struct CF_BRIDGED_TYPE(NSData) __CFData * CFDataRef;
      |                                                         ^
In file included from Sample_objc.m:2:
In file included from ./Sample_objc.h:6:
In file included from ./Sample.h:4:
../cqlrt_cf/cqlrt_cf.h:26:15: error: typedef redefinition with different types ('void *' vs 'const struct __CFString *')
   26 | typedef void *CFStringRef;
      |               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFBase.h:510:61: note: previous definition is here
  510 | typedef const struct CF_BRIDGED_TYPE(NSString) __CFString * CFStringRef;
      |                                                             ^
In file included from Sample_objc.m:2:
In file included from ./Sample_objc.h:6:
In file included from ./Sample.h:4:
../cqlrt_cf/cqlrt_cf.h:27:16: error: typedef redefinition with different types ('int8_t' (aka 'signed char') vs 'unsigned char')
   27 | typedef int8_t Boolean;
      |                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/MacTypes.h:311:41: note: previous definition is here
  311 | typedef unsigned char                   Boolean;
      |                                         ^
Sample_objc.m:6:17: warning: class implementation may not have super class
    6 | @implementation CGSChildRT : NSObject {
      |                 ^
Sample_objc.m:7:25: error: instance variable is already declared
    7 |   CGSChildRS *_Nullable _resultSet;
      |                         ^
./Sample_objc.h:28:25: note: previous definition is here
   28 |   CGSChildRS *_Nullable _resultSet;
      |                         ^
Sample_objc.m:16:17: warning: class implementation may not have super class
   16 | @implementation CGSChildRS : NSObject {
      |                 ^
Sample_objc.m:17:24: error: instance variable is already declared
   17 |   Child_result_set_ref _resultSet;
      |                        ^
./Sample_objc.h:13:24: note: previous definition is here
   13 |   Child_result_set_ref _resultSet;
      |                        ^
Sample_objc.m:62:17: warning: class implementation may not have super class
   62 | @implementation CGSOutArgThingRT : NSObject {
      |                 ^
```

**Reply by ricomariani** [reply]
So in "legacy" mode you have to declare the instance variables in the header AND in the implementation, so I had to do that to get it to compile on my system but this is obviously not going to work.  And it seems there are other issues too.  It was a lot to ask for it to work at all given that I couldn't even compile it in ARC mode...

However, one thing you have to do is remove the kludge flag:   -DCQL_OBJC_MIN_COMPILE 

I had to add this to avoid the dependence on  CoreFoundation.h which I don't even have.

That will get rid of some of the errors.

If you look for @interface in the python in both places the instance variables between "{" and "}" should be deleted.

That should give you "what's really broken" which maybe I can fix... But I won't be able to build it anymore.  Maybe I can put the extra instance variables under #ifdef CQL_OBJC_MIN_COMPILE so it still at least builds the dumb way.  

It's more important to make sure that the legacy simpler objc output still works though.  `cqlrt_cf/cqlobjc.py` is the replacement for `--rt objc` and it (as far as I can tell) makes byte for byte the same output as before except some helpers are in a different order.  You should be able to use that instead of `--rt objc` (i.e. make the json with `--rt json_schema` first then run the python)

Alternately you can edit it to your hearts content in your own project, it's pretty approachable.

**Reply by ricomariani** [reply]
I've made a bunch of changes in make.sh to remove the various workarounds I had to get a local build.  
There is a --legacy flag that generates the extra instance variable decls needed on Linux, it's off by default now
Make.sh does not set CQL_OBJC_MIN_COMPILE which will get remove the type conflicts

This means it has a hope of building regularly with no changes

**Reply by ricomariani** [reply]
OK I borrowed a Mac and fixed the last few things.

```
$ make.sh
building cql
building C code
building JSON
building OBJC header (.h)
building OBJC implementation (.m)
compiling sample consumer
running demo
dumping result set: count = 5
2025-01-07 11:52:13.075 demo[18389:336655] Row 0: name:name_0 blob:blob_0 age:0 thing:0 key1:code_10 key2:code_20
2025-01-07 11:52:13.076 demo[18389:336655] Row 1: name:name_1 blob:blob_1 age:1 thing:1 key1:code_11 key2:code_21
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 0: irow:1 t:'1'
2025-01-07 11:52:13.076 demo[18389:336655] Row 2: name:name_2 blob:blob_2 age:2 thing:2 key1:code_12 key2:code_22
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 0: irow:1 t:'1'
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 1: irow:2 t:'2'
2025-01-07 11:52:13.076 demo[18389:336655] Row 3: name:name_3 blob:blob_3 age:3 thing:3 key1:code_13 key2:code_23
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 0: irow:1 t:'1'
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 1: irow:2 t:'2'
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 2: irow:3 t:'3'
2025-01-07 11:52:13.076 demo[18389:336655] Row 4: name:name_4 blob:blob_4 age:4 thing:4 key1:code_14 key2:code_24
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 0: irow:1 t:'1'
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 1: irow:2 t:'2'
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 2: irow:3 t:'3'
2025-01-07 11:52:13.076 demo[18389:336655]     Child Row 3: irow:4 t:'4'
```

I also verified that cqlrt_cf still works with the new objc generator.
